src/arch/cuda/hybridAPI/hapi_impl.cpp

```diff
@@ -519,6 +689,173 @@ hapiWorkRequest::hapiWorkRequest() :
+  }
+
+  if (!CsvAccess(gpu_manager).shm_name.empty()) {
+    shm_unlink(CsvAccess(gpu_manager).shm_name.c_str());
+    CsvAccess(gpu_manager).shm_name.clear();
+  }
+}
+
+// Create CUDA IPC handles and populate shared memory region
+// Invoked by all PEs
+static void ipcHandleCreate() {
+  // Only device reps should continue to perform the following operations
+  // so that they are done only once per device
+  if (!CpvAccess(device_rep)) return;
+
+  // Create CUDA IPC memory handle in shared memory
+  DeviceManager& my_dm = CsvAccess(gpu_manager).device_managers[CpvAccess(my_device)];
+  auto comm_buffer = my_dm.get_comm_buffer();
+  CmiAssert(comm_buffer);
+  cudaIpcMemHandle_t* shm_mem_handle = (cudaIpcMemHandle_t*)((char*)CsvAccess(gpu_manager).shm_my_ptr +
+      CsvAccess(gpu_manager).shm_chunk_size * CpvAccess(my_device));
+  void* device_ptr = comm_buffer->base_ptr;
+  hapiCheck(cudaIpcGetMemHandle(shm_mem_handle, device_ptr));
+
+  // Create CUDA IPC events and store them locally (in cuda_ipc_device_info),
+  // and create corresponding IPC handles in shared memory
+  cuda_ipc_device_info& my_device_info = CsvAccess(gpu_manager).cuda_ipc_device_infos[my_dm.global_index];
+  cuda_ipc_event_shared* shm_event_shared = (cuda_ipc_event_shared*)((char*)shm_mem_handle + sizeof(cudaIpcMemHandle_t));
+
+  for (int i = 0; i < CsvAccess(gpu_manager).cuda_ipc_event_pool_size; i++) {
+    cuda_ipc_event_shared* cur_shm_event_shared = shm_event_shared + i;
+
+    my_device_info.event_pool_flags.push_back(0);
+    my_device_info.event_pool_buff_offsets.push_back(0);
+    my_device_info.src_event_pool.emplace_back();
+    my_device_info.dst_event_pool.emplace_back();
+    hapiCheck(cudaEventCreateWithFlags(&my_device_info.src_event_pool[i],
+          cudaEventDisableTiming | cudaEventInterprocess));
+    hapiCheck(cudaEventCreateWithFlags(&my_device_info.dst_event_pool[i],
+          cudaEventDisableTiming | cudaEventInterprocess));
+    hapiCheck(cudaIpcGetEventHandle(&cur_shm_event_shared->src_event_handle,
+          my_device_info.src_event_pool[i]));
+    hapiCheck(cudaIpcGetEventHandle(&cur_shm_event_shared->dst_event_handle,
+          my_device_info.dst_event_pool[i]));
+  }
+
+  // Store device comm buffer ptr in local info (just in case)
+  my_device_info.buffer = device_ptr;
+}
+
+// Open CUDA IPC handles created by other processes
+// Invoked by PE rank 0 of each process
+static void ipcHandleOpen() {
+  // Loop through all processes on this host
+  for (int i = 0; i < CmiNumNodes() / CmiNumPhysicalNodes(); i++) {
+    if (i == CmiMyNodeRankLocal()) continue;
+
+    // Loop through GPU devices per process
+    for (int j = 0; j < CsvAccess(gpu_manager).device_count; j++) {
+      int device_index = CsvAccess(gpu_manager).device_count * i + j;
+      cuda_ipc_device_info& cur_device_info = CsvAccess(gpu_manager).cuda_ipc_device_infos[device_index];
+
+      // Open memory handle
+      cudaIpcMemHandle_t* shm_mem_handle =
```

Comment:
If I'm understanding correctly, is the POSIX shm stuff setup for cuda IPC?

Reply:
Yes, POSIX shared memory is used to exchange CUDA IPC handles between processes on the same physical node.

Reply:
Okay, sounds good.